src/schemas/appSettings.ts

```diff
@@ -0,0 +1,71 @@
+import { z } from 'zod'
+
+export const ProxySettingsSchema = z
+  .object({
+    mode: z.enum(['regular', 'upstream']),
+    port: z
+      .number({ message: 'Port number is required' })
+      .int()
+      .min(1)
+      .max(65535, { message: 'Port number must be between 1 and 65535' }),
+    findPort: z.boolean(),
+    upstream: z.object({
+      url: z.string().url({ message: 'Invalid URL' }).or(z.literal('')),
+      requireAuth: z.boolean(),
+      username: z.string().optional(),
+      password: z.string().optional(),
+    }),
+  })
```

Comment:
Not sure about storing the upstream URL in settings if the mode is set to `regular'. Would a discriminated union work better here?
TS example:
```ts
interface RegularModeSettings {
  mode: 'regular'
  port: number
  findPort: boolean
}

interface UpstreamModeSettings {
  mode: 'upstream'
  port: number
  findPort: boolean
  url: string
  requireAuth: boolean
  username: string
  password: string
}

type ProxySettings = RegularModeSettings | UpstreamModeSettings
```

There

Reply:
I don't think we plan on making app settings shareable, so it's probably not _too important_, but I think of a setting file as a representation of the current app settings, so having stuff that doesn't affect the app there seems odd to me.

Remembering previous choices may or may not be nice, depending on the setting IMO. For example, if I switch between two different configs often, then it makes a lot of sense, but if it's a setting I change once a year or so, then I'd find it confusing.

Reply:
> For example, if I switch between two different configs often, then it makes a lot of sense

Although in this case, maybe the UX would need to be different: let the user create multiple proxy configurations and add a way to choose the active one in the recorder

Reply:
I ended up getting a `discriminatedUnion` to work. Now we're not persistent upstream configuration if proxy mode is regular. Same with the browser path when it's checked to detect it automatically!